### PR TITLE
Fix plugins-browser-list placeholders

### LINKED ISSUE #56793: [Bug]: Searching for plugins in Calypso doesn't have pagination
### Quick summary

We will need to add pagination to the Calypso plugins search list. Let's use the Marketplace new designs for plugins search list oU6nEIeiKSg8CayhEtgR6b-fi-6559%3A46946

### Steps to reproduce

1. Start at `https://wordpress.com/plugins/[DOMAIN]`
2. Search for `woocommerce`

### What you expected to happen

I expected to be able to find more than 24 plugins

### What actually happened

SS 2021-10-05 at 18 55 50

### Context

_No response_

### Browser

_No response_

### Simple/Atomic

_No response_

### Other notes

_No response_

### Reproducibility

Consistent

### Severity

_No response_

### Available workarounds?

_No response_

### Workaround details

_No response_

### PR DESCRIPTION
**Pull request by gcsecsey**
#### Changes proposed in this Pull Request

* Refactored the logic of showing placeholders to cover all cases correctly
* Added a required prop to the test mock

#### Testing instructions

* Go to Calypso live
* Click on the "See All" button on the top right of the "Popular" or "New" sections and do a sanity check, the initial render should show the placeholder as before
* Scroll down and validate that the placeholders are being appended to the plugins list
* Search for a plugin
* Do a sanity check, the initial render should show the placeholder as before
* Navigate to other pages with the pagination control, and validate that the placeholders are displayed in place of the whole list

| Event, `props` | Before | After |
| ----- | ------ | ----- |
| Initial render (works as before), `plugins = [], showPlaceholders = false, paginated = false` | calypso localhost_3000_plugins_popular_gcsecseyfree wordpress com | calypso localhost_3000_plugins_popular_gcsecseyfree wordpress com |
| Initial render (works as before), `plugins = [], showPlaceholders = false, paginated = true` | calypso localhost_3000_plugins_popular_gcsecseyfree wordpress com_s=woocommerce |  calypso localhost_3000_plugins_popular_gcsecseyfree wordpress com_s=woocommerce |
| Requesting next page, `plugins = [ {...}, {...}, {...} ], showPlaceholders = true, paginated = true` | wordpress com_plugins_gcsecseyfree wordpress com_s=woocommerce (1) |  calypso localhost_3000_plugins_popular_gcsecseyfree wordpress com_s=woocommerce (1) |
| Requesting next page, `plugins = [ {...}, {...}, {...} ], showPlaceholders = true, paginated = false` | wordpress com_plugins_gcsecseyfree wordpress com_s=woocommerce (2) | calypso localhost_3000_plugins_popular_gcsecseyfree wordpress com (1) |

Fixes #56793

### CODE REVIEW

File: client/my-sites/plugins/plugins-browser-list/index.jsx [hunk 1]
```diff
@@ -50,29 +48,24 @@ class PluginsBrowserList extends Component {
...
+		if ( showPlaceholders ) {
+			if ( ! paginated ) {
+				// this case is needed to handle infinite scroll
+				return this.renderPluginsViewList().concat( this.renderPlaceholdersViews() );
+			}
 			return this.renderPlaceholdersViews();
+		} else if ( plugins.length ) {
+			return this.renderPluginsViewList();
 		}
+		return this.renderPlaceholdersViews();
```
Comment (cpapazoglou) [inline client/my-sites/plugins/plugins-browser-list/index.jsx]: Can we simplify the logic here by avoiding nested if statements? Also, it feels weird that there is a `return this.renderPlaceholdersViews();` outside the the `if ( showPlaceholders ) {` condition.

**Review by cpapazoglou — CHANGES_REQUESTED** [review]
Thanks for revisiting this @gcsecsey! 

I tested this following your testing instructions and works well. I just found an instance where it doesn't work as expected though:

1. Go to `/plugin/:siteId`
2. Click "Engagement"
3. Go back or click "Plugins" from the sidebar

You will notice that "Featured" section will show some placeholder below the results already shown.

SS 2021-10-13 at 12 20 16

Also, I have added a comment about simplifying the logic.

**Comment by gcsecsey** [comment]
The e2e pipeline failed on one of the editor tests, so I've tried restarting it manually.

File: client/my-sites/plugins/plugins-browser-list/index.jsx (hunk 1 above)
Comment (gcsecsey) [reply client/my-sites/plugins/plugins-browser-list/index.jsx]: I've managed to simplify the logic a bit by introducing the `variant` prop and creating different variants for the different loading methods, please take a look at the new changes! 🙂

**Review by cpapazoglou — COMMENTED** [review]
Just FYI, we prefer rebasing instead of merging 

> Make sure your branch merges cleanly and consider rebasing against trunk to keep the branch history short and clean.

https://github.com/Automattic/wp-calypso/blob/c90a58a5cc567aadf3bda8bd4072b466bad2ff38/docs/CONTRIBUTING.md#L88

File: client/my-sites/plugins/plugins-browser-list/README.md [hunk 2]
```diff
@@ -31,4 +31,10 @@ export default localize( MyPluginsList );
...
 - `site`: a string containing the slug of the selected site
 - `addPlaceholders`: if present, indicates that there should placeholders inserted after the real components list
-- `paginated`: the component can be used with pagination or infinite scroll, if set to true, it will replace all the rendered items with placeholders while loading
+- `variant`: the component can be used with pagination or infinite scroll. This prop controls how to show placeholders when fetching additional data. Defaults to `Fixed`.
+
+  | variant        | behavior                                                                  |
+  | -------------- | ------------------------------------------------------------------------- |
+  | Fixed          | Only shows the data after the initial fetch                               |
+  | Paginated      | Replaces the list with placeholders while the next page is loading        |
+  | InfiniteScroll | Appends placeholder at the end of the list while the next page is loading |
```
Comment (cpapazoglou) [inline client/my-sites/plugins/plugins-browser-list/README.md:40]: Great job!

File: client/my-sites/plugins/plugins-browser-list/index.jsx [hunk 3]
```diff
@@ -50,29 +53,34 @@ class PluginsBrowserList extends Component {
...
+			case PluginsBrowserListVariant.Paginated:
+				if ( showPlaceholders ) {
+					return this.renderPlaceholdersViews();
+				}
+				return this.renderPluginsViewList();
+			case PluginsBrowserListVariant.Fixed:
+			default:
+				return this.renderPluginsViewList();
+		}
 	}
```
Comment (cpapazoglou) [inline client/my-sites/plugins/plugins-browser-list/index.jsx:84]: Thanks for simplifying the logic! It's far better than before 🎉 !

**Review by cpapazoglou — APPROVED** [review]
Tested it and works well under all circumstances!
Thanks for adding tests too!

LGTM, 🚢 !

File: client/my-sites/plugins/plugins-browser-list/README.md (hunk 2 above)
Comment (gcsecsey) [reply client/my-sites/plugins/plugins-browser-list/README.md:40]: Thanks! :)

**Pull request merged**